layout: Let min-content size break before each unbreakable segment run

RichardTjokroutomo opened the pull request:
We were only allowing a soft wrap opportunity before the first run in a `TextRunSegment`, but not before the other runs. For example, this meant that the min-content size of `中文中文` would be 4ic instead of just 1ic.

The change brings the logic in `ContentSizesComputation::process_item()` closer to `TextRunSegment::layout_into_line_items()`.

Testing: Adding new tests. There is also a new failure because we don't support `word-break: auto-phrase`.
Fixes: #39728

RichardTjokroutomo committed:
layout: Fixed add_inline_size() under /layout/flow/inline/mod.rs to compute the proper min_content.

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

Loirooriol reviewed (changes requested):
This assumes that we can always break lines between text runs, which is not right, e.g.

```html
<!DOCTYPE html>
<div style="border: solid; width: min-content">foo<span>bar</span></div>
```

RichardTjokroutomo committed:
layout: Fixed code for PR # 39830. Reverted add_inline_size() back to original and instead calls line_break_opportunity() when there's a line breaking opportunity

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
layout: Fixed accidental typo

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
layout: Fixed accidental typo again (apologies for this, I'm doing this on two devices so there are typos when copying my code from one device to another)

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
layout: Fixed changes from previous commits.

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
layout: Fixed minor typo

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
layout: yet another minor fix.

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
layout: Removed trailing whitespace and ran ./mach fmt to tidy things up

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
cargo.toml: Accidentally modified it in the previous commit, reverting to upstream's version

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
cargo.lock: accidentally removed this file in previous commit

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

Loirooriol reviewed (commented):
I'm not super familiar with the LineBreaker, I will review in more detail later.

RichardTjokroutomo committed:
/ports/servoshell/cargo.toml: Removed the line crown = 0.3.1

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
layout: introduced is_soft_hyphen_slice() in flow/inline/mod.rs.

This is introduced since previous commit failed a single WPT test that involves the CSS property hyphen (issue 39879).

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
layout: Specified the return value of is_soft_hyphen_slice

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo commented:
As a side note, the ```is_soft_hyphen_slice()``` is added for handling one failing WPT test (expected pass). However, given that Servo hasn't implemented the ```hyphens``` property yet, I believe that we should've failed this test. I made relevant [issue](https://github.com/servo/servo/issues/39879) for it.

yezhizhen reviewed (commented):
Just some coding-style suggestion. Clippy is also complaining but following is even better.

RichardTjokroutomo committed:
layout: Minor fix to comply with coding convention

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo committed:
Ran ./mach fmt

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

Loirooriol commented:
> I think a cleaner fix would be to check if `width: min-content`

No, the min-content size shouldn't require that. There are other ways to trigger it.

RichardTjokroutomo commented:
> No, the min-content size shouldn't require that. There are other ways to trigger it.

I see.

RichardTjokroutomo committed:
layout: Ditched linebreaker in favor of manual checking for Chinese characters. I believe that this will be a safer method as I am not familiar with the library used by linebreaker

Signed-off-by: Richard Tjokroutomo <richard.tjokro2@gmail.com>

RichardTjokroutomo commented:
@Loirooriol Got it. So do we revert to previous commit, or?

Loirooriol committed:
Try to align with `TextRunSegment::layout_into_line_items()`

Signed-off-by: Oriol Brufau <obrufau@igalia.com>

Loirooriol committed:
Add tests

Signed-off-by: Oriol Brufau <obrufau@igalia.com>

jdm commented on tests/wpt/meta/css/css-text/white-space/reference/white-space-intrinsic-size-022-ref.html:
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
```
Wrong directory. These files need to be under tests/wpt/tests instead.

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#55470) with upstreamable changes.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#55470).

Loirooriol commented:
I'm not sure what you did, I'm reverting to 405b92d1e539ab37c1e5379260a81965046f0e53.

No need to run the LineBreaker again.
